### ISSUE TITLE
SIGSEGV cause database corruption

### ISSUE THREAD

**Issue by swingbit** [MonetDB/MonetDB#7410]
**Describe the bug**
It happens a bit too often that a SIGSEGV crash then makes the database unusable.

Unfortunately this is not simple to reproduce, but no matter the cause of the crash, the database should not get corrupt.

I can exclude "disk full" as a possible cause.

From this log excerpt:

- a SIGSEGV happens
- the database is restarted but crashed again at start
- the database is restarted but crashed again at start, this time showing that it is corrupt beyond repair

```
2023-10-27 10:48:19 MSG merovingian[8]: database 'db01' (442387) has crashed with signal SIGSEGV (dumped core)
2023-10-27 10:48:19 MSG merovingian[8]: database 'db01' has crashed after start on 2023-10-27 09:36:42, attempting restart, up min/avg/max: 0s/0s/0s, crash average: 1.00 0.10 0.03 (1-0=1)
2023-10-27 10:48:19 MSG db01[456921]: arguments: /opt/monetdb/bin/mserver5 --dbpath=/var/lib/monetdb/dbfarm/db01 --set=merovingian_uri=mapi:monetdb://1ae0b3e9fec4:50000/db01 --set=mapi_listenaddr=none --set=mapi_usock=/var/lib/monetdb/dbfarm/db01/.mapi.sock --set=monet_vault_key=/var/lib/monetdb/dbfarm/db01/.vaultkey --set=gdk_nr_threads=8 --set=tablet_threads=8 --set=max_clients=64 --set=sql_optimizer=sequential_pipe --set=embedded_py=3 --set=mal_for_all=yes
2023-10-27 10:48:19 MSG db01[456921]: # MonetDB 5 server v11.47.12
2023-10-27 10:48:19 MSG db01[456921]: # This is an unreleased version
2023-10-27 10:48:19 MSG db01[456921]: # Serving database 'db01', using 8 threads
2023-10-27 10:48:19 MSG db01[456921]: # Compiled for x86_64-pc-linux-gnu/64bit with 128bit integers
2023-10-27 10:48:19 MSG db01[456921]: # Found 125.898 GiB available main-memory of which we use 102.607 GiB
2023-10-27 10:48:19 MSG db01[456921]: # Copyright (c) 1993 - July 2008 CWI.
2023-10-27 10:48:19 MSG db01[456921]: # Copyright (c) August 2008 - 2023 MonetDB B.V., all rights reserved
2023-10-27 10:48:19 MSG db01[456921]: # Visit https://www.monetdb.org/ for further information
2023-10-27 10:48:20 MSG db01[456921]: # MonetDB/GIS module loaded
2023-10-27 10:48:20 MSG db01[456921]: # MonetDB/Python3 module loaded
2023-10-27 10:48:27 MSG db01[456921]: # MonetDB/SQL module loaded
2023-10-27 10:48:27 MSG db01[456921]: # Listening for UNIX domain connection requests on mapi:monetdb:///var/lib/monetdb/dbfarm/db01/.mapi.sock
2023-10-27 10:48:32 MSG merovingian[8]: database 'db01' (456921) has crashed with signal SIGSEGV (dumped core)
2023-10-27 10:48:36 MSG merovingian[8]: database 'db01' has crashed after start on 2023-10-27 10:48:19, attempting restart, up min/avg/max: 0s/0s/0s, crash average: 1.00 0.20 0.07 (2-0=2)
2023-10-27 10:48:36 MSG db01[456940]: arguments: /opt/monetdb/bin/mserver5 --dbpath=/var/lib/monetdb/dbfarm/db01 --set=merovingian_uri=mapi:monetdb://1ae0b3e9fec4:50000/db01 --set=mapi_listenaddr=none --set=mapi_usock=/var/lib/monetdb/dbfarm/db01/.mapi.sock --set=monet_vault_key=/var/lib/monetdb/dbfarm/db01/.vaultkey --set=gdk_nr_threads=8 --set=tablet_threads=8 --set=max_clients=64 --set=sql_optimizer=sequential_pipe --set=embedded_py=3 --set=mal_for_all=yes
2023-10-27 10:48:37 MSG db01[456940]: # MonetDB 5 server v11.47.12
2023-10-27 10:48:37 MSG db01[456940]: # This is an unreleased version
2023-10-27 10:48:37 MSG db01[456940]: # Serving database 'db01', using 8 threads
2023-10-27 10:48:37 MSG db01[456940]: # Compiled for x86_64-pc-linux-gnu/64bit with 128bit integers
2023-10-27 10:48:37 MSG db01[456940]: # Found 125.898 GiB available main-memory of which we use 102.607 GiB
2023-10-27 10:48:37 MSG db01[456940]: # Copyright (c) 1993 - July 2008 CWI.
2023-10-27 10:48:37 MSG db01[456940]: # Copyright (c) August 2008 - 2023 MonetDB B.V., all rights reserved
2023-10-27 10:48:37 MSG db01[456940]: # Visit https://www.monetdb.org/ for further information
2023-10-27 10:48:37 MSG db01[456940]: # MonetDB/GIS module loaded
2023-10-27 10:48:37 MSG db01[456940]: # MonetDB/Python3 module loaded
2023-10-27 10:48:37 MSG db01[456940]: #main thread: cleanup_and_swap: WARNING: Failed to set bat(6246) transient
2023-10-27 10:48:37 MSG db01[456940]: #main thread: cleanup_and_swap: WARNING: Failed to set bat(6250) transient
2023-10-27 10:48:37 MSG db01[456940]: #main thread: cleanup_and_swap: WARNING: Failed to set bat(6247) transient
2023-10-27 10:48:37 MSG db01[456940]: #main thread: cleanup_and_swap: WARNING: Failed to set bat(6248) transient
2023-10-27 10:48:37 MSG db01[456940]: #main thread: cleanup_and_swap: WARNING: Failed to set bat(6249) transient
2023-10-27 10:48:37 MSG db01[456940]: #main thread: cleanup_and_swap: WARNING: Failed to set bat(6251) transient
2023-10-27 10:48:37 ERR db01[456940]: #main thread: decref: ERROR: BBPrelease: tmp_14146 does not have logical references.
2023-10-27 10:48:37 ERR db01[456940]: #main thread: log_find_bat: ERROR: logger_find_bat failed to find bid for object 22078
2023-10-27 10:48:37 ERR db01[456940]: #main thread: store_load: CRITICAL: Cannot load catalog tables
2023-10-27 10:48:41 MSG merovingian[8]: database 'db01' (456940) has crashed with signal SIGSEGV (dumped core)
```

**Software versions**
 - MonetDB version number 11.47.12
 - OS and version: Fedora 38
 - self-installed and compiled

**Comment by njnes** [comment]
with the cause of the SEGV found, can we close this bug (although not sure not if a bug or the segv caused the corruption)?

**Comment by njnes** [comment]
It seems the 'corruption' has been resolved by https://www.monetdb.org/hg/MonetDB/rev/36890da7c473.

**Comment by swingbit** [comment]
The issue seems now completely solved. Here's a summary:
- The SEGV turned out to be rather irrelevant. It was caused by my own code, but it was not triggering any corruption. It was simply stopping the database.
- The corruption occurred when restarting the database (with or without SEGV). This has been fixed in 56f26e28c24a83118a609f3a5315579795fa9d51
- Further investigation with this database led to another important fix about processing commits, where it could happen that the catalog would reference bats that were however still only in the WAL. The fix for this is in 29fa610c246a11fb2c056658acf24a79ccef1826

I will close this.
